File: sdk/ts-sdk/src/marketplace.ts
/**
 * Abyss Marketplace operations
 */

import { DemiurgeClient } from './client';
import { Listing, Address } from './types';
import { cgtToSmallest } from './utils';

export class MarketplaceApi {
  constructor(private client: DemiurgeClient) {}

  /**
   * Get all active listings
   */
  async getAllListings(): Promise<Listing[]> {
    return this.client.call<Listing[]>('abyss_getAllListings');
  }

  /**
   * Get a specific listing by ID
   */
  async getListing(listingId: number): Promise<Listing> {
    return this.client.call<Listing>('abyss_getListing', { id: listingId });
  }

  /**
   * Create a new listing (Archon-only)
   */
  async createListing(
    nftId: number,
    seller: Address,
    price: number | string,
    signedTxHex: string
  ): Promise<Listing> {
    const priceStr = typeof price === 'number' ? cgtToSmallest(price) : price;

    return this.client.call<Listing>('abyss_createListing', {
      nft_id: nftId,
      seller,
      price: priceStr,
      signed_tx_hex: signedTxHex,
    });
  }

  /**
   * Cancel a listing
   */
  async cancelListing(
    listingId: number,
    seller: Address,
    signedTxHex: string
  ): Promise<void> {
    await this.client.call('abyss_cancelListing', {
      id: listingId,
      seller,
      signed_tx_hex: signedTxHex,
    });
  }

  /**
   * Buy a listing
   */
  async buyListing(
    listingId: number,
    buyer: Address,
    signedTxHex: string
  ): Promise<void> {
    await this.client.call('abyss_buyListing', {
      id: listingId,
      buyer,
      signed_tx_hex: signedTxHex,
    });
  }
}

File: sdk/rust-sdk/src/marketplace.rs
/*!
 * Abyss Marketplace operations
 */

use crate::client::DemiurgeClient;
use crate::error::Result;
use crate::types::Listing;
use serde_json::json;

pub struct MarketplaceApi<'a> {
    client: &'a DemiurgeClient,
}

impl<'a> MarketplaceApi<'a> {
    pub fn new(client: &'a DemiurgeClient) -> Self {
        Self { client }
    }

    /// Get all active listings
    pub async fn get_all_listings(&self) -> Result<Vec<Listing>> {
        self.client.call("abyss_getAllListings", None).await
    }

    /// Get a specific listing by ID
    pub async fn get_listing(&self, listing_id: u64) -> Result<Listing> {
        self.client
            .call("abyss_getListing", Some(json!({ "id": listing_id })))
            .await
    }
}

